ruby/ql/lib/codeql/ruby/frameworks/ActiveRecord.qll

```diff
@@ -516,3 +516,175 @@ private module Persistence {
+ *
+ * ```rb
+ * class User
+ *   has_many :posts
+ *   has_one :profile
+ * end
+ * ```
+ */
+private class ActiveRecordAssociation extends DataFlow::CallNode {
+  private ActiveRecordModelClass modelClass;
+
+  ActiveRecordAssociation() {
+    not exists(this.asExpr().getExpr().getEnclosingMethod()) and
+    this.asExpr().getExpr().getEnclosingModule() = modelClass and
+    this.getMethodName() = ["has_one", "has_many", "belongs_to", "has_and_belongs_to_many"]
+  }
+
+  /**
+   * Gets the class which declares this association.
+   *  For example, in
+   *  ```rb
+   *  class User
+   *    has_many :posts
+   *  end
+   *  ```
+   *  the source class is `User`.
+   */
+  ActiveRecordModelClass getSourceClass() { result = modelClass }
+
+  /**
+   * Gets the class which this association refers to.
+   *  For example, in
+   *  ```rb
+   *  class User
+   *    has_many :posts
+   *  end
+   *  ```
+   *  the target class is `Post`.
+   */
+  ActiveRecordModelClass getTargetClass() {
+    result.getName().toLowerCase() = this.getTargetModelName()
+  }
+
+  /**
+   * Gets the (lowercase) name of the model this association targets.
+   * For example, in `has_many :posts`, this is `post`.
+   */
+  string getTargetModelName() {
+    exists(string s |
+      s = this.getArgument(0).asExpr().getExpr().getConstantValue().getStringlikeValue()
+    |
+      // has_one :profile
+      // belongs_to :user
+      this.isSingular() and
+      result = s
+      or
+      // has_many :posts
+      // has_many :stories
+      this.isCollection() and
+      pluralize(result) = s
+    )
+  }
+
+  /** Holds if this association is one-to-one */
+  predicate isSingular() { this.getMethodName() = ["has_one", "belongs_to"] }
+
+  /** Holds if this association is one-to-many or many-to-many */
+  predicate isCollection() { this.getMethodName() = ["has_many", "has_and_belongs_to_many"] }
+}
+
+/**
+ * Converts `input` to plural form.
+ */
+bindingset[input]
+bindingset[result]
+private string pluralize(string input) {
+  exists(string stem | stem + "y" = input | result = stem + "ies")
+  or
+  result = input + "s"
+}
+
+/**
+ * A call to a method generated by an ActiveRecord association.
+ * These yield ActiveRecord collection proxies, which act like collections but
+ * add some additional methods.
+ * We exclude `<model>_changed?` and `<model>_previously_changed?` because these
+ * do not yield ActiveRecord instances.
+ * https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html
+ */
+private class ActiveRecordAssociationMethodCall extends DataFlow::CallNode {
+  ActiveRecordAssociation assoc;
+
+  ActiveRecordAssociationMethodCall() {
+    exists(string model | model = assoc.getTargetModelName() |
+      this.getReceiver().(ActiveRecordInstance).getClass() = assoc.getSourceClass() and
+      (
+        assoc.isCollection() and
+        (
+          this.getMethodName() = pluralize(model) + ["", "=", "<<"]
```

Comment:
It's not clear to me from the tests that the `others<<` method is registering as a an `ActiveRecordAssociationMethodCall` in practice - in the test file on line 29 we have:
```
author2.posts << post2
```
but the `.expected` only has a single result
```
| associations.rb:29:1:29:13 | call to posts | associations.rb:5:1:9:3 | Post |
```
for the call to `posts` - I would expect another one for the call to `posts<<`.

I think that we might parse this as a call to a method named `<<` where the receiver is the result of a call to `posts`? If this is the case, then moving `<<` to `ActiveRecordCollectionProxyMethodCall` should fix this. 

`=` is probably worth having a test case for as well, although I think that one does work as it's written here.